### DISCUSSION TITLE
How would you use Autofac for the DI?

### DISCUSSION THREAD

**Discussion by dmoonfire** [statiqdev/Discussions#68] [Q&A] [answered]
I have a preference of using Autofac over Microsoft's DependencyInjection because of the registration tooling. I see some places that hint as how they could be implemented but I'm not sure what would be the best way to get Autofac injected into the system. Could you please give me some pointers?

Thank you.

**Comment by daveaglick** [comment, marked as answer]
Great question! The Statiq internals are (mostly) designed around the Microsoft.Extensions.DependencyInjection abstraction layer and don't do anything very exotic beyond lifetime management. There's a few places like the Razor module where the container is cracked open a bit and the service descriptors are messed with, but even that should be okay since it's an abstraction over the container implementation (I.e. Autofac should be providing implementations of all the service descriptor bits).

I got curious and did a search for everywhere an actual `Microsoft.Extensions.DependencyInjection.ServiceCollection` implementation is instantiated and was happy to see past @daveaglick had already through through your scenario! There's two points at which a `IServiceCollection` is created:
- The `Bootstrapper` virtual `CreateServiceCollection()` method (which returns null by default, at which point a new `ServiceCollection` instance will be created for null returns).
- The `Engine` when an `IServiceCollection` is not provided in it's constructor.

I'm going to guess you're using the bootstrapper? If so, the trick is to derive your own bootstrapper class from `Bootstrapper` and override `CreateServiceCollection()`. To use your own bootstrapper you'll need to bypass the `BootstrapperFactory` convenience class, but the only downside is slightly longer syntax.

To use Statiq Web with your own bootstrapper that uses a different DI container you might end up with something like this:

```
new AutofacBootstrapper(args)
    .AddDefaults()
    .AddWeb()
    //...
```

Let me know if you have any trouble!

**Reply by dmoonfire** [reply]
Thank you.
